File: rust/arete-a4-sdk/src/wallet.rs
//! Wallet adapter boundary for the Arete SDK.
//!
//! Rust port of `typescript/core/src/wallet/types.ts`. The core SDK is
//! intentionally RPC-free: it only constructs
//! [`BuiltInstruction`](crate::instruction::BuiltInstruction) values.
//! Everything network-related (recent blockhash, message compilation, signing,
//! sending, and confirmation) lives behind the [`WalletAdapter`] boundary,
//! implemented by adapters that wrap the Solana library of your choice
//! (`solana-sdk` keypairs for scripts, remote signers, etc.).
//!
//! Divergences from the TypeScript surface (by design):
//!
//! - TS wallet failures are arbitrary thrown values that the executor
//!   duck-types (`normalizeTransactionError` /
//!   `getTransactionFailureOutcome` walk `cause` chains looking for outcome
//!   shapes, 4001 rejection codes, and program error codes). Rust adapters
//!   instead classify their own failures: [`WalletError`] carries an optional
//!   structured
//!   [`TransactionFailureOutcome`](crate::operations::TransactionFailureOutcome)
//!   which the executor consumes directly. A `WalletError` without an outcome
//!   is classified as not-submitted in the send phase.
//! - The optional TS `inspectTransaction` capability is not part of the trait
//!   yet; it arrives with the transaction-transport integration pass.

use std::error::Error as StdError;
use std::fmt;
use std::sync::Arc;

use crate::instruction::BuiltInstruction;
use crate::operations::TransactionFailureOutcome;
use crate::transactions::TransactionTransport;

/// Confirmation level for transaction processing.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum ConfirmationLevel {
    /// Transaction processed but not confirmed.
    Processed,
    /// Transaction confirmed by the cluster.
    Confirmed,
    /// Transaction finalized (recommended for production).
    Finalized,
}

impl ConfirmationLevel {
    /// Wire/display name (`"processed" | "confirmed" | "finalized"`).
    pub fn as_str(&self) -> &'static str {
        match self {
            ConfirmationLevel::Processed => "processed",
            ConfirmationLevel::Confirmed => "confirmed",
            ConfirmationLevel::Finalized => "finalized",
        }
    }
}

impl fmt::Display for ConfirmationLevel {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(self.as_str())
    }
}

/// Options forwarded to the wallet adapter when sending a transaction.
///
/// The core SDK does not interpret these; it passes them straight through to
/// the adapter, which owns all RPC semantics. `extra` is the Rust rendering of
/// the TS index signature: adapter-specific passthrough options (priority
/// fees, lookup tables, etc.).
#[derive(Debug, Clone, Default)]
pub struct SendOptions {
    /// Confirmation level the adapter should wait for.
    pub confirmation_level: Option<ConfirmationLevel>,
    /// Skip the RPC preflight simulation.
    pub skip_preflight: Option<bool>,
    /// Adapter-specific passthrough options.
    pub extra: serde_json::Map<String, serde_json::Value>,
}

/// Result returned by a wallet adapter after broadcasting a transaction.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SendResult {
    /// Transaction signature (base58).
    pub signature: String,
    /// Slot in which the transaction landed, if the adapter reports it.
    pub slot: Option<u64>,
}

/// Execution context passed through to wallet adapters.
///
/// Mirror of the TS `WalletExecutionContext`: the client passes its
/// [`TransactionTransport`] on every `sign_and_send` so adapters can fetch
/// blockhashes, simulate, send, and poll signature status through the stack
/// relay instead of a direct RPC connection. `#[non_exhaustive]` keeps future
/// fields non-breaking; construct via [`WalletExecutionContext::new`] or
/// `Default`.
#[non_exhaustive]
#[derive(Clone, Default)]
pub struct WalletExecutionContext {
    /// Transaction relay transport supplied by the executing client, if any.
    pub transaction_transport: Option<Arc<dyn TransactionTransport>>,
}

impl WalletExecutionContext {
    /// Context carrying the given transaction transport.
    pub fn new(transaction_transport: Option<Arc<dyn TransactionTransport>>) -> Self {
        Self {
            transaction_transport,
        }
    }
}

impl fmt::Debug for WalletExecutionContext {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("WalletExecutionContext")
            .field(
                "transaction_transport",
                &self.transaction_transport.as_ref().map(|_| "TransactionTransport"),
            )
            .finish()
    }
}

/// Failure reported by a wallet adapter.
///
/// Adapters classify their own failures: when the adapter knows how far the
/// transaction got (wallet rejection, submitted-but-unconfirmed, failed on
/// chain with a program error code), it attaches a structured
/// [`TransactionFailureOutcome`] via [`WalletError::with_outcome`] and the
/// operation executor consumes it directly (no TS-style duck-typing of thrown
/// values). Without an outcome the executor classifies the failure as
/// not-submitted in the send phase.
#[derive(Debug, thiserror::Error)]
#[error("{message}")]
pub struct WalletError {
    message: String,
    outcome: Option<TransactionFailureOutcome>,
    #[source]
    source: Option<Box<dyn StdError + Send + Sync + 'static>>,
}

impl WalletError {
    /// Create a wallet error from a plain message.
    pub fn new(message: impl Into<String>) -> Self {
        WalletError {
            message: message.into(),
            outcome: None,
            source: None,
        }
    }

    /// Create a wallet error whose message is derived from a structured
    /// failure outcome.
    pub fn from_outcome(outcome: TransactionFailureOutcome) -> Self {
        WalletError {
            message: outcome.message().to_string(),
            outcome: Some(outcome),
            source: None,
        }
    }

    /// Attach a structured failure outcome (how far the transaction got).
    pub fn with_outcome(mut self, outcome: TransactionFailureOutcome) -> Self {
        self.outcome = Some(outcome);
        self
    }

    /// Attach the underlying error source (RPC error, IO error, etc.).
    pub fn with_source(
        mut self,
        source: impl Into<Box<dyn StdError + Send + Sync + 'static>>,
    ) -> Self {
        self.source = Some(source.into());
        self
    }

    /// Human-readable failure message.
    pub fn message(&self) -> &str {
        &self.message
    }

    /// The structured failure outcome, if the adapter classified one.
    pub fn outcome(&self) -> Option<&TransactionFailureOutcome> {
        self.outcome.as_ref()
    }

    /// Consume the error, yielding its structured outcome or a
    /// not-submitted fallback at `fallback_phase` carrying the message.
    pub fn into_outcome(
        self,
        fallback_phase: crate::operations::FailurePhase,
    ) -> TransactionFailureOutcome {
        match self.outcome {
            Some(outcome) => outcome,
            None => TransactionFailureOutcome::NotSubmitted {
                phase: fallback_phase,
                message: self.message,
            },
        }
    }
}

impl From<String> for WalletError {
    fn from(message: String) -> Self {
        WalletError::new(message)
    }
}

impl From<&str> for WalletError {
    fn from(message: &str) -> Self {
        WalletError::new(message)
    }
}

/// Wallet adapter interface for signing and sending transactions.
///
/// Implementations own blockhash fetching, message compilation (legacy or
/// v0), signing, sending, and confirmation. The core SDK only needs
/// [`public_key`](WalletAdapter::public_key) for signer-account resolution and
/// [`sign_and_send`](WalletAdapter::sign_and_send) to broadcast built
/// instructions.
#[async_trait::async_trait]
pub trait WalletAdapter: Send + Sync {
    /// The wallet's public key as a base58-encoded string.
    fn public_key(&self) -> String;

    /// Signer addresses the adapter can satisfy without per-send signers.
    ///
    /// Defaults to `[self.public_key()]`.
    fn signer_addresses(&self) -> Vec<String> {
        vec![self.public_key()]
    }

    /// Compile, sign, and broadcast one or more built instructions as a
    /// single transaction.
    ///
    /// Accepting a slice (rather than a single instruction) makes batching
    /// and composition fall out for free.
    async fn sign_and_send(
        &self,
        instructions: &[BuiltInstruction],
        options: &SendOptions,
        context: &WalletExecutionContext,
    ) -> Result<SendResult, WalletError>;
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::operations::FailurePhase;

    struct FixedKeyWallet;

    #[async_trait::async_trait]
    impl WalletAdapter for FixedKeyWallet {
        fn public_key(&self) -> String {
            "wallet-address".to_string()
        }

        async fn sign_and_send(
            &self,
            _instructions: &[BuiltInstruction],
            _options: &SendOptions,
            _context: &WalletExecutionContext,
        ) -> Result<SendResult, WalletError> {
            Err(WalletError::new("unused"))
        }
    }

    #[test]
    fn signer_addresses_default_to_public_key() {
        let wallet = FixedKeyWallet;
        assert_eq!(wallet.signer_addresses(), vec!["wallet-address".to_string()]);
    }

    #[test]
    fn wallet_error_without_outcome_falls_back_to_phase() {
        let error = WalletError::new("connection reset");
        assert_eq!(error.message(), "connection reset");
        assert!(error.outcome().is_none());
        assert_eq!(
            error.into_outcome(FailurePhase::Send),
            TransactionFailureOutcome::NotSubmitted {
                phase: FailurePhase::Send,
                message: "connection reset".to_string(),
            }
        );
    }

    #[test]
    fn wallet_error_prefers_attached_outcome() {
        let outcome = TransactionFailureOutcome::SubmittedUnknown {
            signature: "sig".to_string(),
            slot: Some(42),
            message: "confirmation timed out".to_string(),
        };
        let error = WalletError::from_outcome(outcome.clone());
        assert_eq!(error.message(), "confirmation timed out");
        assert_eq!(error.into_outcome(FailurePhase::Wallet), outcome);
    }

    #[test]
    fn wallet_error_carries_source() {
        let io = std::io::Error::other("socket closed");
        let error = WalletError::new("send failed").with_source(io);
        let source = std::error::Error::source(&error).expect("source");
        assert_eq!(source.to_string(), "socket closed");
    }
}

File: python/arete-sdk/arete/wallet.py
"""Wallet adapter boundary for the Arete SDK.

Python projection of ``typescript/core/src/wallet/types.ts`` with the Rust
SDK's classified-failure model (``rust/arete-a4-sdk/src/wallet.rs``).

The core SDK is intentionally RPC-free: it only constructs
:class:`arete.instructions.BuiltInstruction` values. Everything
network-related (recent blockhash, message compilation, signing, sending, and
confirmation) lives behind the :class:`WalletAdapter` boundary, implemented by
adapters that wrap the Solana library of your choice (a raw keypair signer for
scripts, a remote signer, ...).

Divergences from the TypeScript surface (idiom, not semantics):

- TS wallet failures are arbitrary thrown values that the executor duck-types
  (``normalizeTransactionError`` walks ``cause`` chains looking for outcome
  shapes and 4001 rejection codes). Python adapters instead classify their own
  failures: :class:`WalletError` carries an optional structured
  :class:`TransactionFailureOutcome` which the operation executor consumes
  directly. A ``WalletError`` without an outcome is classified as
  ``not-submitted`` in the ``send`` phase.
- The outcome model (four terminal statuses ``confirmed | not-submitted |
  submitted-unknown | chain-failed``, each with the phase that produced it)
  lives here because the wallet boundary is the classifier; ``arete.operations``
  re-exports it.
- Program errors inside outcomes reuse
  :class:`arete.instructions.ErrorMetadata` (``code``/``name``/``msg``) rather
  than a separate ``ProgramError`` shape.
"""

from __future__ import annotations

from dataclasses import dataclass, field, replace
from typing import (
    Any,
    Mapping,
    Optional,
    Protocol,
    Sequence,
    Tuple,
    runtime_checkable,
)

from arete.errors import AreteError
from arete.instructions import (
    BuiltInstruction,
    ErrorMetadata,
    format_program_error,
)
from arete.transactions import TransactionTransport

CONFIRMATION_LEVELS: Tuple[str, ...] = ("processed", "confirmed", "finalized")

FAILURE_STATUSES: Tuple[str, ...] = (
    "not-submitted",
    "submitted-unknown",
    "chain-failed",
)

FAILURE_PHASES: Tuple[str, ...] = (
    "build",
    "wallet",
    "send",
    "confirmation",
    "chain",
)

_PHASES_BY_STATUS = {
    "not-submitted": ("build", "wallet", "send"),
    "submitted-unknown": ("send", "confirmation"),
    "chain-failed": ("confirmation", "chain"),
}

_SEND_OPTION_FIELDS = ("confirmation_level", "skip_preflight", "signers")


@dataclass(frozen=True)
class SendOptions:
    """Options forwarded to the wallet adapter when sending a transaction.

    The core SDK does not interpret these; it passes them straight through to
    the adapter, which owns all RPC semantics. ``extra`` is the Python
    rendering of the TS index signature: adapter-specific passthrough options
    (priority fees, lookup tables, ...). ``signers`` are optional extra local
    signers for this send; their concrete type depends on the adapter.
    """

    confirmation_level: Optional[str] = None
    skip_preflight: Optional[bool] = None
    signers: Optional[Tuple[Any, ...]] = None
    extra: Mapping[str, Any] = field(default_factory=dict)

    def __post_init__(self) -> None:
        if (
            self.confirmation_level is not None
            and self.confirmation_level not in CONFIRMATION_LEVELS
        ):
            raise ValueError(
                f"confirmation_level must be one of {CONFIRMATION_LEVELS}, "
                f"got {self.confirmation_level!r}"
            )
        if self.signers is not None and not isinstance(self.signers, tuple):
            object.__setattr__(self, "signers", tuple(self.signers))

    @classmethod
    def coerce(cls, value: Any) -> "SendOptions":
        """``None`` → defaults; :class:`SendOptions` passthrough; a mapping's
        unknown keys land in ``extra`` (adapter passthrough)."""
        if value is None:
            return cls()
        if isinstance(value, cls):
            return value
        if isinstance(value, Mapping):
            known = {name: value[name] for name in _SEND_OPTION_FIELDS if name in value}
            extra = {
                key: item
                for key, item in value.items()
                if key not in _SEND_OPTION_FIELDS and key != "extra"
            }
            nested = value.get("extra")
            if isinstance(nested, Mapping):
                extra.update(nested)
            return cls(extra=extra, **known)
        raise TypeError(
            f"send options must be a SendOptions or mapping, got {type(value).__name__}"
        )

    def merged(self, overrides: Optional["SendOptions"]) -> "SendOptions":
        """Field-wise merge where ``overrides`` wins; ``extra`` maps merge."""
        if overrides is None:
            return self
        return SendOptions(
            confirmation_level=(
                overrides.confirmation_level
                if overrides.confirmation_level is not None
                else self.confirmation_level
            ),
            skip_preflight=(
                overrides.skip_preflight
                if overrides.skip_preflight is not None
                else self.skip_preflight
            ),
            signers=overrides.signers if overrides.signers is not None else self.signers,
            extra={**dict(self.extra), **dict(overrides.extra)},
        )

    def with_signers(self, signers: Optional[Sequence[Any]]) -> "SendOptions":
        if signers is None:
            return self
        return replace(self, signers=tuple(signers))


@dataclass(frozen=True)
class SendResult:
    """Result returned by a wallet adapter after broadcasting a transaction."""

    signature: str
    slot: Optional[int] = None


@dataclass(frozen=True)
class WalletExecutionContext:
    """Execution context passed through to wallet adapters.

    The executing client passes its :class:`arete.transactions.TransactionTransport`
    on every ``sign_and_send`` so adapters can fetch blockhashes, simulate,
    send, and poll signature status through the stack relay instead of a
    direct RPC connection.
    """

    transaction_transport: Optional[TransactionTransport] = None


@dataclass(frozen=True)
class TransactionInspectionResult:
    """Unsigned transaction inspection returned by a capable wallet adapter.

    Inspection must not sign or submit the transaction.
    """

    fee_lamports: Optional[int] = None
    logs: Optional[Tuple[str, ...]] = None
    compute_units_consumed: Optional[int] = None
    context_slot: Optional[int] = None
    error: Any = None
    extra: Mapping[str, Any] = field(default_factory=dict)


# ---------------------------------------------------------------------------
# Transaction outcome model (canonical §7: outcomes are data)
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class ConfirmedTransactionOutcome:
    """The confirmed terminal status (normally reported through receipts)."""

    signature: str
    slot: Optional[int] = None

    status: str = "confirmed"
    phase: str = "confirmation"


@dataclass(frozen=True)
class TransactionFailureOutcome:
    """One of the three failure terminal statuses with the phase that
    produced it. Outcomes are data — the executor raises
    :class:`arete.operations.OperationExecutionError` *holding* one of these.

    Prefer the :meth:`not_submitted` / :meth:`submitted_unknown` /
    :meth:`chain_failed` factories, which validate status/phase combinations
    and derive a default message.
    """

    status: str
    phase: str
    message: str = ""
    signature: Optional[str] = None
    slot: Optional[int] = None
    program_error: Optional[ErrorMetadata] = None
    cause: Any = None

    def __post_init__(self) -> None:
        if self.status not in FAILURE_STATUSES:
            raise ValueError(
                f"status must be one of {FAILURE_STATUSES}, got {self.status!r}"
            )
        allowed = _PHASES_BY_STATUS[self.status]
        if self.phase not in allowed:
            raise ValueError(
                f"phase for status '{self.status}' must be one of {allowed}, "
                f"got {self.phase!r}"
            )
        if self.status == "submitted-unknown" and not self.signature:
            raise ValueError("submitted-unknown outcomes require a signature")
        if not self.message:
            object.__setattr__(self, "message", _derive_outcome_message(self))

    @classmethod
    def not_submitted(
        cls,
        phase: str = "send",
        *,
        message: str = "",
        cause: Any = None,
    ) -> "TransactionFailureOutcome":
        return cls(status="not-submitted", phase=phase, message=message, cause=cause)

    @classmethod
    def submitted_unknown(
        cls,
        signature: str,
        *,
        phase: str = "confirmation",
        slot: Optional[int] = None,
        message: str = "",
        cause: Any = None,
    ) -> "TransactionFailureOutcome":
        return cls(
            status="submitted-unknown",
            phase=phase,
            signature=signature,
            slot=slot,
            message=message,
            cause=cause,
        )

    @classmethod
    def chain_failed(
        cls,
        *,
        phase: str = "chain",
        signature: Optional[str] = None,
        slot: Optional[int] = None,
        program_error: Optional[ErrorMetadata] = None,
        message: str = "",
        cause: Any = None,
    ) -> "TransactionFailureOutcome":
        return cls(
            status="chain-failed",
            phase=phase,
            signature=signature,
            slot=slot,
            program_error=program_error,
            message=message,
            cause=cause,
        )


def _default_outcome_message(outcome: "TransactionFailureOutcome") -> str:
    if outcome.status == "not-submitted":
        return f"Transaction was not submitted during {outcome.phase}"
    if outcome.status == "submitted-unknown":
        return (
            f"Transaction {outcome.signature} was submitted but its status is unknown"
        )
    if outcome.program_error is not None:
        return format_program_error(outcome.program_error)
    if outcome.signature:
        return f"Transaction {outcome.signature} failed on chain"
    return "Transaction failed on chain"


def _derive_outcome_message(outcome: "TransactionFailureOutcome") -> str:
    cause = outcome.cause
    if isinstance(cause, BaseException) and str(cause):
        return str(cause)
    return _default_outcome_message(outcome)


class WalletError(AreteError):
    """Failure reported by a wallet adapter.

    Adapters classify their own failures: when the adapter knows how far the
    transaction got (wallet rejection, submitted-but-unconfirmed, failed on
    chain with a program error code), it attaches a structured
    :class:`TransactionFailureOutcome`; the operation executor consumes it
    directly. Without an outcome the executor classifies the failure as
    ``not-submitted`` in the ``send`` phase.
    """

    def __init__(
        self,
        message: str = "",
        *,
        outcome: Optional[TransactionFailureOutcome] = None,
        cause: Any = None,
    ) -> None:
        if not message and outcome is not None:
            message = outcome.message
        super().__init__(message or "Wallet operation failed", "WALLET_ERROR")
        self.outcome = outcome
        self.cause = cause if cause is not None else (
            outcome.cause if outcome is not None else None
        )

    @classmethod
    def from_outcome(cls, outcome: TransactionFailureOutcome) -> "WalletError":
        return cls(outcome.message, outcome=outcome)

    def into_outcome(self, fallback_phase: str = "send") -> TransactionFailureOutcome:
        """The attached structured outcome, or a not-submitted fallback at
        ``fallback_phase`` carrying this error's message."""
        if self.outcome is not None:
            return self.outcome
        return TransactionFailureOutcome.not_submitted(
            fallback_phase, message=self.message, cause=self.cause or self
        )


@runtime_checkable
class WalletAdapter(Protocol):
    """Wallet adapter interface for signing and sending transactions.

    Implementations own blockhash fetching, message compilation (legacy or
    v0), signing, sending, and confirmation. The core SDK only needs
    ``public_key`` for signer-account resolution and ``sign_and_send`` to
    broadcast built instructions.

    Optional capabilities (checked structurally with ``getattr``):

    - ``signer_addresses`` — addresses the adapter can satisfy without
      per-send signers (defaults to ``[public_key]``).
    - ``async def inspect_transaction(instructions, options=None, context=None)
      -> TransactionInspectionResult`` — unsigned inspection; must not sign,
      submit, or prompt a wallet.

    Failures raise :class:`WalletError` (classified when possible).
    """

    public_key: str

    async def sign_and_send(
        self,
        instructions: Sequence[BuiltInstruction],
        options: Optional[SendOptions] = None,
        context: Optional[WalletExecutionContext] = None,
    ) -> SendResult:
        """Compile, sign, and broadcast built instructions as one transaction.

        Accepting a sequence (rather than a single instruction) makes batching
        and composition fall out for free.
        """
        ...


def wallet_signer_addresses(wallet: Any) -> Tuple[str, ...]:
    """Signer addresses an adapter can satisfy, including its public key."""
    if wallet is None:
        return ()
    addresses = []
    declared = getattr(wallet, "signer_addresses", None)
    if declared is not None and not callable(declared):
        addresses.extend(str(address) for address in declared)
    public_key = getattr(wallet, "public_key", None)
    if isinstance(public_key, str) and public_key:
        addresses.append(public_key)
    seen = set()
    unique = []
    for address in addresses:
        if address not in seen:
            seen.add(address)
            unique.append(address)
    return tuple(unique)
